File: osint_nexus/core/detectors/tls.py
from collections.abc import Mapping
from beartype import beartype

from osint_nexus.core.db.fingerprint_repository import FingerprintRepository
from osint_nexus.core.detectors.base import FingerprintStrategy
from osint_nexus.core.type_defs import JSONValue


class TlsFingerprintStrategy(FingerprintStrategy[Mapping[str, JSONValue], Mapping[str, JSONValue]]):
    """Strategy for TLS (JA3) fingerprinting."""

    name: str = "tls_ja3"

    def __init__(self, repo: FingerprintRepository | None = None) -> None:
        self.repo = repo or FingerprintRepository()

    @beartype
    def extract(self, data: Mapping[str, JSONValue]) -> Mapping[str, JSONValue]:
        # Expecting data to be a dictionary containing the ja3 hash string
        ja3_hash = str(data.get("ja3_hash", "unknown"))

        device_info = self.repo.get_signature("ja3", ja3_hash) or "unknown"

        return {
            "name": self.name,
            "data": {
                "ja3_hash": ja3_hash,
                "inferred_device": device_info,
            },
            "confidence": 0.90 if device_info != "unknown" else 0.10,
        }

File: internal/engine/strategies/tls.go
package strategies

import (
	"context"

	"github.com/FJ-cyberzilla/osint-nexus/internal/db"
	"github.com/FJ-cyberzilla/osint-nexus/internal/types"
)

// TLSPayload implements types.FingerprintPayload for TLS fingerprinting input.
type TLSPayload struct {
	JA3Hash string `json:"ja3_hash"`
}

// PayloadType returns the payload type identifier.
func (p TLSPayload) PayloadType() string { return "tls_input" }

// TLSOutputPayload implements types.FingerprintPayload for TLS fingerprinting output.
type TLSOutputPayload struct {
	JA3Hash        string `json:"ja3_hash"`
	InferredDevice string `json:"inferred_device"`
}

// PayloadType returns the payload type identifier.
func (p TLSOutputPayload) PayloadType() string { return "tls_output" }

// TLSStrategy implements the FingerprintStrategy for TLS (JA3).
type TLSStrategy struct {
	repo *db.FingerprintRepository
}

// NewTLSStrategy initializes a new TLSStrategy with a given FingerprintRepository.
func NewTLSStrategy(repo *db.FingerprintRepository) *TLSStrategy {
	return &TLSStrategy{
		repo: repo,
	}
}

// Name returns the strategy name.
func (s *TLSStrategy) Name() string {
	return "tls_ja3"
}

// Extract performs JA3 fingerprinting.
func (s *TLSStrategy) Extract(ctx context.Context, data types.FingerprintData) (types.FingerprintResult, error) {
	payload, ok := data.Payload.(TLSPayload)
	if !ok {
		return types.FingerprintResult{
			Name: s.Name(),
			Data: types.FingerprintData{
				Type:    "tls_ja3",
				Payload: TLSOutputPayload{JA3Hash: "unknown", InferredDevice: "unknown"},
			},
			Confidence: 0.0,
		}, nil
	}

	deviceInfo := s.repo.GetSignature("ja3", payload.JA3Hash)
	if deviceInfo == "" {
		deviceInfo = "unknown"
	}

	confidence := 0.10
	if deviceInfo != "unknown" {
		confidence = 0.90
	}

	return types.FingerprintResult{
		Name: s.Name(),
		Data: types.FingerprintData{
			Type:    "tls_ja3",
			Payload: TLSOutputPayload{JA3Hash: payload.JA3Hash, InferredDevice: deviceInfo},
		},
		Confidence: confidence,
	}, nil
}
